JESSICA BELTRAN

SPRINT 8

In [1]:
import pandas as pd
import numpy as np
import re


NIVELL 1
Exercici 1:  Exploració inicial del dataset

🎯 Objectiu: entendre l’estructura general del dataset i detectar possibles problemes abans de començar a netejar o transformar dades.
En aquest exercici has de realitzar una exploració inicial del DataFrame per respondre, com a mínim, a les següents preguntes:

» Quantes files i columnes té el dataset?
» Quin tipus de dades conté cada columna?
» Hi ha files duplicades?
» Les dades semblen coherents a primera vista?
» Detectes possibles problemes o inconsistències evidents?
Per fer-ho, hauràs de:

» Analitzar les dimensions del dataset
» Mostrar una mostra representativa de files
» Revisar els tipus de variables
» Les dades semblen coherents a primera vista?
» Detectar files duplicades
» Identificar i descriure possibles problemes inicials (sense solucionar-los encara)

In [2]:
# LLlegim l'arxiu i li afegim el header =3 ja que es on es troba la capçalera correcta de dataframe.
# li diem que ens mostri les 5 primeres files  i ja el tenim mes net 

df = pd.read_excel('sprint8_complex.xlsx', header =3)
df.head(5)

,Unnamed: 0,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1


In [3]:
# El dataframe té 1007 files i 20 columnes

df.shape

(1007, 20)

In [4]:
# Mètode que ens permet veure, valors nuls, tipus de variables, buits, etc. Ens dona una idea de com són les dades i quines columnes hi ha.

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  1007 non-null   int64  
 1   Nom                         1006 non-null   str    
 2   Cognoms                     1006 non-null   str    
 3   DNI                         1006 non-null   str    
 4   País d'origen               1006 non-null   str    
 5   Ciutat                      1006 non-null   str    
 6   Dia de Naixement            1006 non-null   float64
 7   Mes de Naixement            1006 non-null   float64
 8   Any de Naixement            1006 non-null   float64
 9   Gènere                      966 non-null    str    
 10  Salari mensual              1006 non-null   str    
 11  Fills                       419 non-null    object 
 12  No Fills                    585 non-null    object 
 13  Grup Professional           1005 non-null   

In [5]:
# Mètode que ens permet veure estadístiques bàsiques de les columnes numèriques del dataframe, com la mitjana, la desviació estàndard, els valors mínims i màxims, i els percentils. Ens dóna una idea de la distribució de les dades.
df.describe()

,Unnamed: 0,Dia de Naixement,Mes de Naixement,Any de Naixement,Nombre_fills,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
count,1007.000000,1006.000000,1006.000000,1006.000000,709.000000,315.000000,261.000000,936.000000
mean,503.000000,15.027833,8.556660,2002.229622,1.203103,55951.050794,11.024521,15.379167
std,290.840162,10.327436,14.298532,143.118077,1.389404,62095.437005,8.775557,7.276425
min,0.000000,-99.000000,0.000000,1940.000000,0.000000,-50.000000,0.500000,-10.000000
25%,251.500000,7.000000,3.250000,1969.000000,0.000000,22632.000000,4.700000,12.800000
50%,503.000000,15.000000,6.000000,1982.000000,1.000000,56135.000000,9.800000,15.500000
75%,754.500000,24.000000,10.000000,1995.000000,2.000000,80000.000000,13.600000,18.300000
max,1006.000000,32.000000,99.000000,3000.000000,5.000000,1000000.000000,25.000000,42.000000


In [6]:
# Veiem 5 files aleatòries per tenir un altre enfoc de les dades i veure si hi ha algun patró o informació rellevant que no es veu en les primeres files.

df.sample(5)

,Unnamed: 0,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
653,653,Laia,Moreno Pujol,65110751A,Espanya,Palma,20.0,4.0,1986.0,H,1117 euros,NaN,NaN,Grup A,Analista,NaN,True,20000.0,8.4,20.3
966,966,Sergio,Ruiz Gómez,32342927V,Espanya,Palma,30.0,8.0,1959.0,H,842 €,NaN,True,Grup A,Responsable de vendes,0.0,False,NaN,NaN,18.5
978,978,Giulia,Conti Russo,72192924S,Ità lia,Bologna,29.0,10.0,1990.0,H,1931 euros,NaN,True,Grup C,Administratiu,0.0,NaN,NaN,NaN,15.0
261,261,Guillem,Soler Roca,17083242X,Espanya,Girona,20.0,5.0,1995.0,H,"2,309 €",True,NaN,Grup C,Analista,3.0,False,NaN,NaN,16.1
675,675,Alejandro,Navarro,50530014N,Espanya,Tarragona,26.0,11.0,1966.0,D,"1,947 €",NaN,True,Grup B,Senior analyst,NaN,True,50000.0,10.7,-5.0


In [7]:
# Mirem si hi han files duplicades, per fer-ho treiem la columna Unnamed: 0 ja que ens desvirtua les dades estan duplicada amb l'índex de pandas

comparacio_columnes = [col for col in df.columns if col != 'Unnamed: 0']
df.duplicated(subset = comparacio_columnes).sum()


np.int64(4)

In [8]:
# Mirem si hi han valors nuls a les columnes del dataframe (ens dona el total per columna).
df.isnull().sum()

Unnamed: 0                      0
Nom                             1
Cognoms                         1
DNI                             1
País d'origen                   1
Ciutat                          1
Dia de Naixement                1
Mes de Naixement                1
Any de Naixement                1
Gènere                         41
Salari mensual                  1
Fills                         588
No Fills                      422
Grup Professional               2
Càrrec                         2
Nombre_fills                  298
Te_cotxe                      349
Km_anuals                     692
Consum_mitja_L_100km          746
Temperatura_mitjana_ciutat     71
dtype: int64

### Conclusions

Després de l'exploració inicial, hem trobat diferents punts a tenir en compte per a la neteja de dades:

- **Columnes sense nom a la capçalera:** hem aplicat el mètode `header=3` de forma provisional, per poder fer l'exploració del DataFrame de manera més eficient.
- **Columna `Unnamed: 0`:** sembla un índex redundant, ja que duplica la informació que Pandas ja proporciona amb el seu propi índex intern.
- **Valors nuls:** hi ha diverses columnes amb valors nuls, en proporcions diferents (des d'1 fins a centenars de registres).
- **Dades no estandarditzades:** per exemple, la columna `Salari mensual` presenta formats inconsistents (separador de milers amb coma, ús indistint de "€" i "euros", amb o sense espai).
- **Caràcters estranys en diferents cel·les:** tant en noms de columnes com en valors de text, probablement per un problema de codificació (encoding) en llegir l'arxiu.
- **4 files completament duplicades**, detectades excloent la columna `Unnamed: 0` de la comparació.
- **Possibles DNIs repetits:** a falta de verificar-ho amb detall (Exercici 2).
- **Tipus de dades incorrectes:** columnes com `Dia/Mes/Any de Naixement` i `Nombre_fills` apareixen com a `float64` en lloc de `int` (per la presència de valors nuls), i `Fills`/`No Fills`/`Te_cotxe` apareixen com a `object` en lloc de `bool`.

Exercici 2: Validacions bàsiques de les dades i decisions d'eliminació

🎯 Objectiu: assegurar que les dades compleixen unes condicions mínimes de qualitat i determinar quines files poden continuar sent analitzades i quines s'han de descartar.
En aquest exercici hauràs de validar aspectes clau del dataset que són crítics en qualsevol context real d'anàlisi de dades, prenent decisions justificades sobre la validesa o eliminació de registres.

Aspectes a validar:

» Existeixen identificadors únics (DNI) per a cada persona?
» Què faràs amb les files duplicades o registres repetits?
» Hi ha valors obligatoris que falten o informació incoherent?
» Detectes formats incorrectes, errors de codificació o valors impossibles?
Concretament, hauràs de:

» Corregir possibles problemes de codificació i accentuació en els noms de les columnes i en els valors de text, si escau.
» Detectar i eliminar files duplicades exactes, indicant quants registres s'han eliminat.
» Comprovar la unicitat del DNI. En cas de detectar DNIs duplicats, hauràs d'analitzar la coherència de la informació associada i descartar o anul·lar els registres que no siguin fiables, justificant clarament la decisió presa.
» Identificar els camps clau que no haurien de contenir valors nuls, detectar registres amb informació insuficient i eliminar únicament aquelles files completament buides o amb informació irrecuperable o no substituïble, explicant per què no és possible ni raonable imputar-les.

### Codificació

Detallem pas a pas els problemes resolts:

- **Problema detectat:** diversos noms de columna i valors de text contenien caràcters estranys (`Muñoz`, `Cà rrec`, `€`...), típic d'un text guardat en UTF-8 però interpretat incorrectament amb una altra codificació en llegir-lo (fenomen conegut com a *mojibake*).
- **Solució aplicada:** es reverteix l'error d'interpretació amb `.encode('cp1252').decode('utf-8')` — es recuperen els bytes originals (assumint que es van llegir malament com a Windows-1252/cp1252) i es tornen a interpretar correctament com a UTF-8. *(S'ha fet servir cp1252 en comptes de Latin-1 perquè alguns caràcters, com el símbol de l'euro mal codificat, només existeixen en el rang de cp1252, no en Latin-1.)*
- **Per què la funció té `isinstance` i `try/except`:**
  - `isinstance(valor, str)` evita aplicar la correcció a valors que no són text (booleans, `NaN`), que no tenen el mètode `.encode()`.
  - El `try/except UnicodeError` evita que el procés s'aturi si algun text està tan malmès que no es pot corregir (per exemple, truncat) — en aquests casos, es deixa el valor tal com està temporalment, sense frenar la resta de la neteja.
- **Tractament dels casos irreparables:** després d'aplicar la correcció, es comprova si encara queda algun rastre de mojibake (patró `Ã`). Si és així, es considera informació irrecuperable i es marca com a valor nul (`NaN`), en comptes de deixar text corromput al dataset.
- **Aplicació:** la funció s'aplica tant als valors de les columnes de text (fent servir `.apply()` en un bucle sobre les columnes `str`) com als propis noms de les columnes (`df.columns`).

In [9]:
# Funció per corregir problemes de codificació (mojibake UTF-8 llegit com Latin-1).
# Si el valor no és text, es retorna tal qual. Si el text és irrecuperable
# (truncat o amb caràcters que no existeixen en Latin-1), també es retorna tal qual,
# gràcies al try/except, en comptes d'aturar tot el procés.

def arreglar_encoding(valor):
    if isinstance(valor, str):
        try:
            return valor.encode('cp1252').decode('utf-8')
        except UnicodeError:
            return valor
    else:
        return valor

# Apliquem la correcció a totes les columnes de tipus text (valors de les cel·les)
columnas_texto = df.select_dtypes(include=['str']).columns
for col in columnas_texto:
    df[col] = df[col].apply(arreglar_encoding)

# Apliquem la mateixa correcció als noms de les columnes

df.columns = [arreglar_encoding(col) for col in df.columns]



In [10]:
# Netejem qualsevol resta de mojibike que no es pugui reparar (marquem com a nul)

for col in df.select_dtypes(include=['str']).columns:
    df[col] = df[col].apply(lambda x: np.nan if isinstance(x, str) and 'Ã' in x else x)

In [11]:
# Comprovació final de les columnes

for col in df.select_dtypes(include=['str']).columns:
    restantes = df[col].astype(str).str.contains('Ã', na=False).sum()
    if restantes > 0:
        print(col, '->', restantes)

In [12]:
df

,Unnamed: 0,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1002,1002,Alba,Muñoz,97954019T,Espanya,Bilbao,25.0,10.0,1983.0,D,"1,409 €",NaN,True,Grup B,Responsable de vendes,0.0,NaN,NaN,NaN,14.7
1003,1003,Claudia,Gil Ferrer,55297384F,Espanya,València,1.0,11.0,1981.0,H,1.789 €,NaN,True,Grup B,Responsable de vendes,0.0,True,29952.0,9.2,20.3
1004,1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1005,1005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Duplicitat de files

- **Problema detectat:** el dataset contenia files completament duplicades (idèntiques en totes les columnes, exceptuant `Unnamed: 0`, que és un índex redundant i s'ha exclòs de la comparació per no falsejar el resultat).
- **Solució aplicada:** s'utilitza `drop_duplicates(subset=...)` per eliminar físicament les files duplicades, conservant la primera aparició de cada grup i descartant la resta.
- **Resultat:** s'han detectat i eliminat **4 files duplicades exactes**.

In [13]:
# Eliminem les files completament duplicades, excloent la columna 'Unnamed: 0'
# (que és un índex redundant i faria que cap fila es considerés mai duplicada)
columnas_comparacio = [col for col in df.columns if col != 'Unnamed: 0']

files_abans = df.shape[0]
df = df.drop_duplicates(subset=columnas_comparacio)
files_despres = df.shape[0]

print(f"S'han eliminat {files_abans - files_despres} files duplicades")

S'han eliminat 4 files duplicades


### DNI's duplicats

- **Problema detectat:** s'han trobat 16 files implicades en DNIs repetits, agrupades en dos patrons diferents:
  - Casos on el DNI es repeteix per a la **mateixa persona** (coincideixen Nom i Any de Naixement), probablement per registres desactualitzats o introduïts més d'un cop.
  - Casos on el mateix DNI apareix associat a **persones clarament diferents** (noms, ciutats i dates de naixement incompatibles) — un error greu, ja que un DNI no pot pertànyer a més d'una persona.
- **Criteri aplicat:**
  - Quan es tracta de la mateixa persona, es conserva únicament la fila amb menys valors nuls (la més completa/fiable), i s'elimina la resta.
  - Quan es tracta de persones diferents amb el mateix DNI, es considera que cap de les files és fiable (no es pot determinar quina és la correcta), i s'eliminen totes les files del grup.
- **Resultat:** s'han eliminat **13 files** per DNI duplicat no fiable.

In [14]:
df['DNI'].duplicated().sum()

np.int64(11)

In [15]:
df[df['DNI'].notna() & df['DNI'].duplicated(keep=False)].sort_values('DNI')

,Unnamed: 0,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
250,250,Laia,Navarro Vila,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,1235 euros,True,True,Grup D,Analista junior,0.0,False,NaN,NaN,10.3
604,604,Laia,Navarro,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,"1,676 €",True,False,Grup A,Analista,1.0,True,50.0,NaN,14.2
654,654,Oriol,Gómez,16618423V,Espanya,Madrid,10.0,1.0,1968.0,Home,1731 euros,True,NaN,Grup B,Data analyst,4.0,NaN,NaN,NaN,NaN
754,754,Oriol,Gómez,16618423V,Espanya,Barcelona,10.0,1.0,1968.0,Home,2.358 €,True,NaN,NaN,NaN,4.0,NaN,NaN,NaN,NaN
758,758,Oriol,Gómez,16618423V,Espanya,Màlaga,10.0,1.0,1968.0,H,1879 euros,True,NaN,Grup B,Data Analyst,4.0,False,NaN,NaN,NaN
759,759,Oriol,Gómez,16618423V,Espanya,Sevilla,10.0,1.0,1968.0,NaN,"2,000 €",NaN,NaN,Grup A,Responsable de vendes,5.0,NaN,NaN,NaN,20.7
558,558,Alexia,NaN,49529438G,Espanya,Màlaga,12.0,10.0,1981.0,Dona,"1,176 €",NaN,True,Grup A,Data Analyst,0.0,False,NaN,NaN,38.0
692,692,Alexia,NaN,49529438G,Espanya,Màlaga,12.0,10.0,1981.0,Dona,1663€,NaN,True,Grup C,Cap de projecte,NaN,True,80000.0,10.7,16.9
89,89,Guillem,Pujol Costa,54013571W,Espanya,A Coruña,17.0,7.0,1983.0,D,1.574 €,True,NaN,Grup B,Project lead,4.0,False,NaN,NaN,14.6
142,142,Marta,Ruiz,54013571W,Espanya,Palma,24.0,5.0,1982.0,H,"1,276 €",True,True,Grup A,Analista,0.0,False,NaN,NaN,15.7


In [16]:
# Guardem el nombre de files abans de resoldre els DNI duplicats
files_abans_dni = df.shape[0]

# Funció per decidir, dins d'un grup amb el mateix DNI, si són la mateixa persona
# (coincideixen Nom i Any de Naixement) o persones diferents.
def resoldre_grup(grup):
    mateixa_persona = (grup['Nom'].nunique() == 1) and (grup['Any de Naixement'].nunique() == 1)
    if mateixa_persona:
        # Ens quedem amb la fila amb menys valors nuls (més informació completa)
        idx_millor = grup.notna().sum(axis=1).idxmax()
        fila = grup.loc[[idx_millor]].copy()
        fila['DNI'] = grup.name  # reconstruïm la columna DNI (Pandas 3.x l'exclou de "grup")
        return fila
    else:
        # Persones diferents amb el mateix DNI: DNI no fiable, eliminem tot el grup
        buida = grup.iloc[0:0].copy()
        buida['DNI'] = []
        return buida

sense_duplicats = df[~df['DNI'].duplicated(keep=False)]
amb_duplicats = df[df['DNI'].duplicated(keep=False)]

resolts = amb_duplicats.groupby('DNI', group_keys=False).apply(resoldre_grup)
df = pd.concat([sense_duplicats, resolts]).sort_index()

files_eliminades_dni = files_abans_dni - df.shape[0]
print(f"S'han eliminat {files_eliminades_dni} files per DNI duplicat no fiable")

S'han eliminat 13 files per DNI duplicat no fiable


In [17]:
df

,Unnamed: 0,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
998,998,Chloé,Dubois,66354268T,França,Nice,2.0,12.0,1958.0,H,"1,954 €",True,NaN,Grup B,Data Analyst,1.0,NaN,NaN,NaN,11.7
999,999,Adrià,NaN,57511543T,Espanya,Palma,31.0,12.0,1991.0,H,"2,010 €",True,NaN,Grup C,Analista,1.0,False,NaN,NaN,-10.0
1001,651,Kasper,Pettersen,62042563B,Noruega,Oslo,11.0,5.0,1996.0,D,2.290 €,NaN,True,Grup B,Senior analyst,0.0,True,80000.0,0.5,6.5
1005,1005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Files buides o amb valors impossibles

- **Problema detectat:** el dataset contenia files amb informació pràcticament inexistent (tots els camps a `NaN`) i altres amb valors clarament impossibles o absurds (any de naixement 3000, mes 99, dia -99, DNI XXXX, país Narnia...) — registres irrecuperables o probablement introduïts com a prova.
- **Solució aplicada:**
  - S'eliminen les files amb menys de 3 valors no nuls d'un total de 20 columnes, ja que no aporten informació analitzable.
  - S'estableixen rangs vàlids per a columnes clau (`Any de Naixement`, `Mes de Naixement`, `Dia de Naixement`, `Nombre_fills`, `Km_anuals`), basats en criteris lògics (per exemple, edat laboral entre 18 i 60 anys). Qualsevol fila amb un valor fora d'aquests rangs es considera no fiable i s'elimina.
- **Resultat:** s'eliminen les files irrecuperables per manca d'informació o per contenir valors impossibles, deixant el dataset net i llest per a l'anàlisi.

In [18]:
# Eliminem les files pràcticament buides (menys de 3 valors no nuls d'un total de 20 columnes)
valors_no_nuls = df.notna().sum(axis=1)
files_abans_buides = df.shape[0]
df = df[valors_no_nuls >= 3]
print(f"S'han eliminat {files_abans_buides - df.shape[0]} files pràcticament buides")

S'han eliminat 1 files pràcticament buides


In [19]:
# Eliminem la columna 'Unnamed: 0', ja que és un índex redundant que duplica
# la informació que Pandas ja proporciona amb el seu propi índex intern.
df = df.drop(columns=['Unnamed: 0'])

In [20]:
rangs_valids = {
    "Any de Naixement": (1940, 2010),
    "Mes de Naixement": (1, 12),
    "Dia de Naixement": (1, 31),
    "Nombre_fills": (0, 10),
    "Km_anuals": (0, 300000),
}

files_abans_rangs = df.shape[0]
mask_valid = pd.Series(True, index=df.index)

for col, (min_val, max_val) in rangs_valids.items():
    fora_de_rang = (df[col] < min_val) | (df[col] > max_val)
    mask_valid &= ~fora_de_rang.fillna(False)

df = df[mask_valid]
print(f"S'han eliminat {files_abans_rangs - df.shape[0]} files amb valors fora de rang (impossibles)")
print(f"Total final de files: {df.shape[0]}")

S'han eliminat 84 files amb valors fora de rang (impossibles)
Total final de files: 905


In [21]:
print(df['Mes de Naixement'].min(), df['Mes de Naixement'].max())
print(df['Dia de Naixement'].min(), df['Dia de Naixement'].max())
print(df['Any de Naixement'].min(), df['Any de Naixement'].max())

1.0 12.0
1.0 31.0
1958.0 2004.0


Exercici 3: Transformacions necessàries i preparació del dataset
🎯 Objectiu: preparar el dataset perquè sigui coherent, analitzable i estructuralment correcte abans de realitzar qualsevol anàlisi descriptiva.
En aquest exercici hauràs d'aplicar transformacions bàsiques però essencials, orientades a millorar la qualitat de les dades i facilitar l'anàlisi posterior.

Aspectes a revisar i transformar:

» Els valors categòrics estan normalitzats i són consistents?
» Els valors nuls són coherents amb la informació disponible?
» Hi ha redundància de variables que pugui generar problemes analítics?
» Les variables contextuals (com temperatura, salari) estan completes i preparades per a l'anàlisi?
Concretament, hauràs de:

» Normalitzar els valors categòrics per evitar ambigüitats (per exemple, valors que representen el mateix concepte però estan escrits de manera diferent).
» Revisar la coherència dels valors nuls, assegurant que:
Les persones que tenen quilometratge anual i consum mitjà associats disposen de cotxe. En els casos en què consti consum mitjà però no quilometratge anual, caldrà decidir com resoldre aquesta incoherència de manera justificada.
Les persones que tenen un valor en el nombre de fills efectivament tenen fills.
» Unificar la informació sobre fills en una única columna, ja que mantenir diverses columnes amb informació complementària no és òptim i pot generar problemes analítics (redundància o multicolinealitat).
» Validar la temperatura mitjana. En cas que aquesta informació falti, hauràs d'imputar-la utilitzant un criteri raonable (per ciutat o per un nivell superior justificat).
» Revisar i imputar els salaris faltants, aplicant criteris analítics coherents (com ara càrrec, grup professional, ciutat, gènere o franja d'edat).

### Normalització de dades

En aquesta secció es revisen les columnes categòriques (gènere, grup professional, càrrec, país d'origen, ciutat) del dataset per detectar valors que representen el mateix concepte però estan escrits de manera diferent (per exemple, "H" i "Home" per referir-se al mateix gènere), unificant-los sota un únic valor per evitar ambigüitats en l'anàlisi posterior.

In [22]:
# Mirem quantes respostes poden donar els usuaris 

df['Gènere'].unique()

<StringArray>
['Dona', 'H', 'D', 'F', 'Home', 'NC', 'M', nan, 'A']
Length: 9, dtype: str

In [23]:
# Comptem quants valors s'han modificat (variants + nuls) abans de la normalització
valors_abans = df['Gènere'].copy()

mapeo_genere = {
    'Dona': 'Dona',
    'D': 'Dona',
    'F': 'Dona',
    'Home': 'Home',
    'H': 'Home',
    'M': 'Home',
    'NC': 'Prefereix no contestar',
    'A': 'Prefereix no contestar',
}

df['Gènere'] = df['Gènere'].map(mapeo_genere).fillna('Prefereix no contestar')

# Comptem quantes files han canviat de valor respecte a l'original
files_modificades = (valors_abans.fillna('NULL_ORIGINAL') != df['Gènere']).sum()
print(f"S'han normalitzat {files_modificades} valors de la columna Gènere")

S'han normalitzat 755 valors de la columna Gènere


In [24]:
# Finalment en deixem nomès tres possibles respostes 

df['Gènere'].unique()

<StringArray>
['Dona', 'Home', 'Prefereix no contestar']
Length: 3, dtype: str

In [25]:
df['Grup Professional'].unique()

<StringArray>
['Grup B', 'Grup C', 'Grup A', 'Grup D']
Length: 4, dtype: str

In [26]:
nuls_abans = df['Grup Professional'].isna().sum()
df['Grup Professional'] = df['Grup Professional'].fillna('Sense grup assignat')
print(f"S'han assignat 'Sense grup assignat' a {nuls_abans} files")

S'han assignat 'Sense grup assignat' a 0 files


In [27]:
df

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,Home,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,Dona,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,Home,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,Laura,Bauer Zimmermann,34637932B,Alemanya,Köln,6.0,3.0,1993.0,Home,"1,644 €",NaN,True,Grup B,Analista,NaN,False,NaN,NaN,11.5
997,Jordi,Hernández,67755039Y,Espanya,Zaragoza,12.0,10.0,1961.0,Dona,"1,074 €",NaN,NaN,Grup A,Data Analyst,1.0,True,80000.0,12.0,12.0
998,Chloé,Dubois,66354268T,França,Nice,2.0,12.0,1958.0,Home,"1,954 €",True,NaN,Grup B,Data Analyst,1.0,NaN,NaN,NaN,11.7
999,Adrià,NaN,57511543T,Espanya,Palma,31.0,12.0,1991.0,Home,"2,010 €",True,NaN,Grup C,Analista,1.0,False,NaN,NaN,-10.0


In [28]:
# Mirem quantes variants apareixen a la columna de càrrec

df['Càrrec'].unique()

<StringArray>
[      'Cap de projecte',        'Senior analyst',             'Tècnic IT',
          'Data Analyst',         'Administratiu', 'Responsable de vendes',
       'Analista junior',          'Project lead',                'Admin.',
               'Analyst',              'Analista',          'Data analyst',
       'Cap de Projecte']
Length: 13, dtype: str

In [29]:

# Normalitzem Càrrec: unifiquem variants d'idioma (anglès/català), majúscules
# i abreviatures que representen el mateix lloc de treball (per exemple,
# "Data Analyst"/"Analyst" i "Analista" es consideren el mateix càrrec).
# Es mantenen com a categories separades els nivells junior/senior, ja que
# representen una diferència real de responsabilitat, no només d'escriptura.
# Els valors nuls s'etiqueten com "Càrrec no especificat" en comptes de deixar-los buits.

carrec_abans = df['Càrrec'].copy()

mapeo_carrec = {
    'Cap de projecte': 'Cap de projecte',
    'Cap de Projecte': 'Cap de projecte',
    'Project lead': 'Cap de projecte',
    'Senior analyst': 'Analista senior',
    'Analyst': 'Analista',
    'Analista': 'Analista',
    'Data Analyst': 'Analista',
    'Data analyst': 'Analista',
    'Analista junior': 'Analista junior',
    'Administratiu': 'Administratiu/va',
    'Admin.': 'Administratiu/va',
    'Tècnic IT': 'Tècnic IT',
    'Responsable de vendes': 'Responsable de vendes',
}

df['Càrrec'] = df['Càrrec'].map(mapeo_carrec).fillna('Càrrec no especificat')

files_modificades = (carrec_abans.fillna('NULL_ORIGINAL') != df['Càrrec']).sum()
print(f"S'han normalitzat {files_modificades} valors de la columna Càrrec")

S'han normalitzat 392 valors de la columna Càrrec


In [30]:
# Comprovem la columna país d'origen per veure les variants i si cal normalitzar (tot correcte)

df["País d'origen"].unique()

<StringArray>
['Espanya', 'Noruega', 'Alemanya', 'Itàlia', 'França']
Length: 5, dtype: str

In [31]:
# Comprovem ciutats per veure variants i si cal normalitzar (tot correcte)

df['Ciutat'].unique()

<StringArray>
[                'Valladolid',                    'Alacant',
                    'Sevilla',                     'Bilbao',
 'Las Palmas de Gran Canaria',                      'Palma',
                  'Barcelona',                  'Stavanger',
                   'Zaragoza',                  'Stuttgart',
                  'Tarragona',                   'València',
                     'Bergen',                     'Milano',
                     'Màlaga',                     'Genova',
                       'Lyon',                      'Paris',
                   'Toulouse',                     'Napoli',
                    'Venezia',                   'A Coruña',
                     'Girona',                       'Oslo',
                     'Murcia',                  'Trondheim',
                     'Madrid',                     'Berlin',
                       'Nice',                      'Lille',
                    'Firenze',                       'Roma',
          

### Coherència Te_cotxe / Km_anuals / Consum_mitjà

- **Problema detectat:** incoherències entre columnes relacionades amb el cotxe — persones amb `Km_anuals` o `Consum_mitjà` registrat però marcades com a `Te_cotxe = False` o buit, i casos amb `Consum_mitjà` però sense `Km_anuals`.
- **Solució aplicada:** s'ha imputat `Km_anuals` amb la mitjana general en els casos on constava `Consum_mitjà` però no `Km_anuals` (2 files), i s'ha corregit `Te_cotxe` a `True` quan hi havia quilometratge registrat (4 files), ja que la presència d'aquest valor implica necessàriament tenir cotxe.

In [32]:
# Comprovem la coherència entre Te_cotxe, Km_anuals i Consum_mitja_L_100km:
# si algú té quilometratge anual o consum mitjà, hauria de tenir cotxe (Te_cotxe = True).
te_km = df['Km_anuals'].notna()
te_consum = df['Consum_mitja_L_100km'].notna()

# Casos on diu que té cotxe però no té ni quilometratge ni consum registrat
print("Te_cotxe=True però sense Km ni Consum:", ((df['Te_cotxe'] == True) & ~te_km & ~te_consum).sum())

# Casos on NO té cotxe (o no consta) però sí té quilometratge registrat -> incoherent
print("Te_cotxe=False/NaN però amb Km:", ((df['Te_cotxe'] != True) & te_km).sum())

# Casos on hi ha consum mitjà però no quilometratge -> incoherència esmentada explícitament a l'enunciat
print("Té Consum però NO té Km:", (te_consum & ~te_km).sum())

# Casos on hi ha quilometratge però no consum -> possible dada incompleta a completar
print("Té Km però NO té Consum:", (te_km & ~te_consum).sum())

Te_cotxe=True però sense Km ni Consum: 14
Te_cotxe=False/NaN però amb Km: 4
Té Consum però NO té Km: 2
Té Km però NO té Consum: 47


In [33]:
# Per als casos on hi ha Consum però no Km, imputem el Km_anuals amb la mitjana
# general de la resta d'empleats que sí tenen aquest valor registrat.
mitjana_km = df['Km_anuals'].mean()

mask_cas3 = te_consum & ~te_km
df.loc[mask_cas3, 'Km_anuals'] = mitjana_km

print(f"S'ha imputat el Km_anuals (mitjana={mitjana_km:.0f}) a {mask_cas3.sum()} files")

S'ha imputat el Km_anuals (mitjana=54533) a 2 files


In [34]:
# Si algú té Km_anuals registrat, per definició ha de tenir cotxe.
# Corregim Te_cotxe a True en aquests casos (la incoherència estava en aquesta columna).
mask_cas2 = (df['Te_cotxe'] != True) & te_km
df.loc[mask_cas2, 'Te_cotxe'] = True

print(f"S'ha corregit Te_cotxe a True en {mask_cas2.sum()} files (tenien Km registrat)")

S'ha corregit Te_cotxe a True en 4 files (tenien Km registrat)


### Unificació de la informació sobre fills

- **Problema detectat:** el dataset mantenia tres columnes redundants i de vegades incoherents entre si (`Fills`, `No Fills`, `Nombre_fills`) — per exemple, files que indicaven tenir fills però amb `Nombre_fills` buit, o al revés.
- **Solució aplicada:** s'unifica tota la informació en una única columna, `Nombre_fills`, ja que és la més específica i completa. Quan `No Fills = True` i no hi havia número registrat, s'imputa 0. La resta de casos ja disposaven del número directament. Finalment, s'eliminen les columnes redundants `Fills` i `No Fills`.
- **Resultat:** 241 files imputades a 0, i el dataset redueix la seva redundància sense perdre informació.

In [35]:
# Comprovem coherència: si Fills=True, hauria d'haver-hi Nombre_fills >= 1
# Si No Fills=True (o Fills=False), Nombre_fills hauria de ser 0
incoherencia_1 = (df['Fills'] == True) & (df['Nombre_fills'].fillna(0) < 1)
incoherencia_2 = (df['No Fills'] == True) & (df['Nombre_fills'].fillna(0) > 0)

print("Fills=True però Nombre_fills < 1:", incoherencia_1.sum())
print("No Fills=True però Nombre_fills > 0:", incoherencia_2.sum())

Fills=True però Nombre_fills < 1: 35
No Fills=True però Nombre_fills > 0: 18


In [36]:
# Unifiquem la informació de fills en una única columna: Nombre_fills.
# Criteri:
# 1. Si Nombre_fills ja té un valor, el mantenim (és la dada més fiable i específica).
# 2. Si Nombre_fills és NaN però No Fills = True, sabem que no té fills -> 0.
# 3. Si Nombre_fills és NaN però Fills = True, sabem que té fills però no quants ->
#    s'imputa amb la mitjana de fills entre les persones que sí en tenen (>0).
# 4. Si no hi ha cap informació a cap de les tres columnes, es queda com a NaN.

mitjana_fills = df.loc[df['Nombre_fills'] > 0, 'Nombre_fills'].mean()

# Cas 2: No Fills = True i Nombre_fills buit -> 0
mask_no_fills = (df['No Fills'] == True) & df['Nombre_fills'].isna()
df.loc[mask_no_fills, 'Nombre_fills'] = 0

# Cas 3: Fills = True i Nombre_fills buit -> mitjana
mask_fills_sense_num = (df['Fills'] == True) & df['Nombre_fills'].isna()
df.loc[mask_fills_sense_num, 'Nombre_fills'] = mitjana_fills

print(f"Casos 'No Fills' sense número, imputats a 0: {mask_no_fills.sum()}")
print(f"Casos 'Fills' sense número, imputats amb mitjana ({mitjana_fills:.1f}): {mask_fills_sense_num.sum()}")

# Eliminem les columnes redundants Fills i No Fills, ja que Nombre_fills ja conté tota la informació
df = df.drop(columns=['Fills', 'No Fills'])

print(f"Columnes finals: {df.shape[1]}")

Casos 'No Fills' sense número, imputats a 0: 241
Casos 'Fills' sense número, imputats amb mitjana (2.1): 0
Columnes finals: 17


### Imputació de la temperatura mitjana

- **Problema detectat:** 64 files sense valor a `Temperatura_mitjana_ciutat`.
- **Solució aplicada:** s'imputa cada valor nul amb la mitjana de temperatura de la mateixa ciutat, ja que és un valor característic i estable per ciutat. Com que cap ciutat tenia tots els seus valors nuls, sempre hi havia una mitjana real disponible per fer la imputació sense necessitat de recórrer a un nivell superior (per exemple, mitjana general o per país).
- **Resultat:** s'han imputat els 64 valors nuls, deixant la columna completa.

In [37]:
print("Nuls a Temperatura_mitjana_ciutat:", df['Temperatura_mitjana_ciutat'].isna().sum())

# Comprovem si hi ha alguna ciutat on TOTS els valors de temperatura són nuls
# (en aquest cas, la mitjana per ciutat no serviria de res, caldria un criteri de nivell superior)
nuls_per_ciutat = df.groupby('Ciutat')['Temperatura_mitjana_ciutat'].apply(lambda x: x.isna().all())
print("Ciutats amb TOTS els valors nuls:", nuls_per_ciutat.sum())

Nuls a Temperatura_mitjana_ciutat: 64
Ciutats amb TOTS els valors nuls: 0


In [38]:
# Imputem la Temperatura_mitjana_ciutat amb la mitjana de la mateixa ciutat,
# ja que és un valor que hauria de ser consistent per a totes les persones
# que viuen a la mateixa ciutat. Com que cap ciutat té tots els valors nuls,
# sempre hi ha una mitjana real disponible per a cada ciutat.

nuls_abans_temp = df['Temperatura_mitjana_ciutat'].isna().sum()

df['Temperatura_mitjana_ciutat'] = df.groupby('Ciutat')['Temperatura_mitjana_ciutat'].transform(
    lambda x: x.fillna(x.mean())
)

print(f"S'han imputat {nuls_abans_temp} valors de Temperatura_mitjana_ciutat (mitjana per ciutat)")

S'han imputat 64 valors de Temperatura_mitjana_ciutat (mitjana per ciutat)


### Neteja i imputació de Salari mensual

- **Problema detectat:** la columna `Salari mensual` era de tipus text, amb múltiples formats inconsistents (separador de milers amb coma o punt, símbol € o paraula "euros", amb o sense espai), i alguns valors escrits directament en paraules (`mil vuit-cents`, `dos mil`, `tres mil`) o sense sentit (`n/d`).
- **Solució aplicada:** es converteix tota la columna a un valor numèric (`float`), traduint els valors en paraules al seu equivalent numèric, eliminant símbols de moneda i separadors de milers, i marcant com a nul els valors irrecuperables (`n/d`). Els salaris que falten s'imputen amb la mitjana salarial del mateix `Càrrec`, ja que persones amb el mateix lloc de treball solen tenir salaris similars.
- **Resultat:** la columna queda completament numèrica i sense valors nuls, llesta per a l'anàlisi.

In [39]:
# Explorem els valors únics de Salari mensual abans de netejar-los,
# per detectar tots els formats diferents presents a la columna.
print("Tipus de dada actual:", df['Salari mensual'].dtype)
print("Nombre de valors únics:", df['Salari mensual'].nunique())
print(df['Salari mensual'].unique()[:40])

# Comprovem si hi ha valors que NO contenen cap dígit (escrits en paraules)
import re
valors_en_paraules = [v for v in df['Salari mensual'].dropna().unique() if not re.search(r'\d', str(v))]
print("\nValors escrits en paraules (sense cap dígit):", valors_en_paraules)

Tipus de dada actual: str
Nombre de valors únics: 778
<StringArray>
[       '1,469 €',        '2,718 €',     '1358 euros',        '1,478 €',
        '1,284 €',        '2,069 €',        '2,349 €',        '2,855 €',
        '1,338 €',          '982 €',        '2,419 €',        '3,672 €',
        '1,003 €', 'mil vuit-cents',        '1,336 €',        '1,385 €',
        '1,751 €',      '889 euros',     '1121 euros',        '1,880 €',
        '1,094 €',          '1479€',        '1,840 €',        '2,660 €',
          '751 €',        '2,628 €',     '2400 euros',          '2564€',
           '927€',        '2,510 €',        '1,538 €',     '3275 euros',
        '1,974 €',        '1,614 €',        '1,014 €',          '2132€',
          '1438€',        '1,923 €',        '1,090 €',        '1,032 €']
Length: 40, dtype: str

Valors escrits en paraules (sense cap dígit): ['mil vuit-cents', 'n/d', 'dos mil', 'tres mil']


In [40]:
# Netegem i convertim Salari mensual a un valor numèric (float).
# Passos:
# 1. Substituïm els valors escrits en paraules (català) pel seu equivalent numèric.
# 2. Marquem com a nul els valors sense sentit ("n/d").
# 3. Eliminem el símbol/paraula de moneda (€, euros) i els espais.
# 4. Eliminem els separadors de milers (comes i punts, mai representen decimals aquí).
# 5. Convertim el resultat a float.

mapeo_paraules = {
    'mil vuit-cents': '1800',
    'dos mil': '2000',
    'tres mil': '3000',
    'n/d': None,
}

def netejar_salari(valor):
    if pd.isna(valor):
        return None
    valor = str(valor).strip()
    if valor in mapeo_paraules:
        valor = mapeo_paraules[valor]
        if valor is None:
            return None
    valor = valor.replace('€', '').replace('euros', '').replace(',', '').replace('.', '').strip()
    try:
        return float(valor)
    except ValueError:
        return None

salaris_abans = df['Salari mensual'].copy()
df['Salari mensual'] = df['Salari mensual'].apply(netejar_salari)

print(df['Salari mensual'].describe())

count     892.000000
mean     1570.606502
std       628.868727
min       673.000000
25%      1085.250000
50%      1415.000000
75%      1899.250000
max      3672.000000
Name: Salari mensual, dtype: float64


In [41]:
# Comprovem si hi ha algun Càrrec sense cap salari vàlid (per assegurar que la mitjana és calculable)
nuls_abans_salari = df['Salari mensual'].isna().sum()

mitjanes_per_carrec = df.groupby('Càrrec')['Salari mensual'].mean()
print(mitjanes_per_carrec)

Càrrec
Administratiu/va         1468.356436
Analista                 1517.312253
Analista junior          1357.709677
Analista senior          1777.137615
Cap de projecte          1767.073394
Responsable de vendes    1566.062500
Tècnic IT                1572.278261
Name: Salari mensual, dtype: float64


In [42]:
# Imputem els salaris faltants amb la mitjana salarial del mateix Càrrec,
# ja que persones amb el mateix lloc de treball solen tenir salaris similars.
# Arrodonim a 2 decimals perquè quedi net com a valor monetari.
df['Salari mensual'] = df['Salari mensual'].fillna(df.groupby('Càrrec')['Salari mensual'].transform('mean'))
df['Salari mensual'] = df['Salari mensual'].round(2)

print(f"S'han imputat {nuls_abans_salari} salaris faltants (mitjana per Càrrec)")
print(df['Salari mensual'].isna().sum())  # hauria de sortir 0

S'han imputat 13 salaris faltants (mitjana per Càrrec)
0


Exercici 4: Primeres anàlisis descriptives
🎯 Objectiu: obtenir una primera visió global del dataset un cop s'han aplicat les transformacions bàsiques, mitjançant estadístiques descriptives i comparacions simples entre grups rellevants.
Aspectes a revisar i analitzar:

» Construir taules resum descriptives.
» Fer comparacions bàsiques entre grups rellevants.
Concretament, hauràs de:

» Construir taules resum descriptives de salaris com una taula resum que permeti veure el sou mig, medià, mínim i màxim per Gènere. Ordena la taula en funció del sou mig.
» Crear una taula resum amb el salari mig per gènere (files) i país d'origen (columnes). Afegeix-hi les mitjanes als marges de la taula. Aplica format condicional a la taula per veure en un color més intens els valors més elevats.
(EXTRA)
Abans d'interpretar mitjanes, medianes o totals, cal preguntar-se si aquests valors permeten una comparació justa entre grups. Quan les mides dels grups són diferents, els percentatges o ràtios solen oferir una visió més objectiva que els valors absoluts.

Podries repetir l’anàlisi utilitzant percentatges i ràtios salarials, i comparar com canvia la interpretació respecte als valors absoluts?
 Important

No es tracta d'extreure conclusions complexes, sinó de descriure què mostren les dades

### Anàlisi descriptiva de salaris

En aquesta secció es construeixen taules resum per obtenir una primera visió global dels salaris del dataset, comparant-los segons diferents variables rellevants (gènere i país d'origen). L'objectiu és descriure patrons bàsics en les dades —com el sou mig, la mediana o els valors extrems per grup— sense encara extreure'n conclusions complexes.

In [43]:
# Taula resum del Salari mensual agrupat per Gènere: mitjana, mediana, mínim i màxim.
# Ordenem per sou mig de major a menor per facilitar la lectura.
taula_genere = df.groupby('Gènere')['Salari mensual'].agg(
    sou_mig='mean',
    sou_media='median',
    sou_min='min',
    sou_max='max'
).sort_values('sou_mig', ascending=False).round(2)

taula_genere

,sou_mig,sou_media,sou_min,sou_max
Gènere,,,,
Home,1701.30,1524.0,726.0,3672.0
Prefereix no contestar,1552.93,1316.0,673.0,3600.0
Dona,1487.66,1380.5,701.0,3564.0


In [44]:
# Taula pivot: salari mig per Gènere (files) i País d'origen (columnes).
# margins=True afegeix una fila i columna amb la mitjana total (marges).
taula_pivot = pd.pivot_table(
    df,
    values='Salari mensual',
    index='Gènere',
    columns="País d'origen",
    aggfunc='mean',
    margins=True,
    margins_name='Mitjana'
).round(2)

taula_pivot

País d'origen,Alemanya,Espanya,França,Itàlia,Noruega,Mitjana
Gènere,,,,,,
Dona,1434.64,1482.63,1412.60,1499.09,1737.17,1487.66
Home,2263.11,1702.69,1675.91,1676.61,1469.06,1701.30
Prefereix no contestar,1819.67,1523.53,1407.14,1767.29,1237.00,1552.93
Mitjana,1649.92,1568.87,1512.83,1590.67,1617.42,1572.04


In [45]:
# Apliquem format condicional: colorem més intens els valors més alts (mapa de color 'YlOrRd')
taula_pivot.style.background_gradient(cmap='YlOrRd').format('{:.2f}')

País d'origen,Alemanya,Espanya,França,Itàlia,Noruega,Mitjana
Gènere,,,,,,
Dona,1434.64,1482.63,1412.60,1499.09,1737.17,1487.66
Home,2263.11,1702.69,1675.91,1676.61,1469.06,1701.30
Prefereix no contestar,1819.67,1523.53,1407.14,1767.29,1237.00,1552.93
Mitjana,1649.92,1568.87,1512.83,1590.67,1617.42,1572.04


### Conclusions

- S'observa que el grup **Home** presenta el sou mig més alt (1.701,30 €), seguit de **Prefereix no contestar** (1.552,93 €) i **Dona** (1.487,66 €). Els sous mínims i màxims són similars entre els tres grups (entre ~670-730 € i ~3.560-3.670 €), la qual cosa suggereix que la diferència de mitjanes no ve donada per valors extrems aïllats, sinó per una tendència general del conjunt de dades.
- **Taula pivot Gènere × País d'origen:** la taula creuada mostra que les diferències salarials no són uniformes entre països: per exemple, el grup **Home** a **Alemanya** presenta la mitjana més elevada de tota la taula (2.263,11 €), molt per sobre de la mitjana general d'Alemanya (1.649,92 €) i de la resta de gèneres al mateix país. En canvi, a **Noruega**, és el grup **Dona** qui presenta la mitjana més alta (1.737,17 €), superant fins i tot als homes noruecs. Això indica que el patró salarial per gènere no es manté constant entre països, i que cal ser prudents abans d'extreure conclusions generals sense tenir en compte el país d'origen.

In [46]:
# Comptem quantes persones hi ha a cada grup de Gènere, per valorar si les mitjanes
# de cada grup són igual de representatives (grups molt petits poden distorsionar la mitjana)
df['Gènere'].value_counts()

Gènere
Dona                      492
Home                      333
Prefereix no contestar     80
Name: count, dtype: int64

In [47]:
# Calculem el salari mig de cada grup com a percentatge respecte a la mitjana general,
# per poder comparar de manera relativa (no només en valor absolut) tenint en compte
# que els grups tenen mides molt diferents.
mitjana_general = df['Salari mensual'].mean()

taula_genere_pct = df.groupby('Gènere')['Salari mensual'].agg(
    n_persones='count',
    sou_mig='mean'
).round(2)

taula_genere_pct['pct_respecte_mitjana'] = (taula_genere_pct['sou_mig'] / mitjana_general * 100).round(1)
taula_genere_pct = taula_genere_pct.sort_values('sou_mig', ascending=False)

taula_genere_pct

,n_persones,sou_mig,pct_respecte_mitjana
Gènere,,,
Home,333,1701.30,108.2
Prefereix no contestar,80,1552.93,98.8
Dona,492,1487.66,94.6


### Conclusions

- Tenint en compte la mida de cada grup (Dona: 492 persones, Home: 333, Prefereix no contestar: 80), s'observa que aquest darrer grup és considerablement més petit, la qual cosa fa que la seva mitjana salarial sigui menys representativa i més sensible a valors extrems individuals.
- Expressant els sous com a percentatge respecte a la mitjana general (1.572,04 €), es confirma la mateixa tendència que amb els valors absoluts: **Home** se situa un 8,2% per sobre de la mitjana, **Dona** un 5,4% per sota, i **Prefereix no contestar** pràcticament en la mitjana (98,8%).
- En aquest cas, la interpretació no canvia substancialment entre valors absoluts i percentatges, ja que els tres grups tenen una mida prou gran (totes per sobre de 80 persones) com perquè les mitjanes siguin raonablement estables. Tot i així, cal recordar que el grup **Prefereix no contestar**, en representar només un 8,8% del total de persones, s'hauria d'interpretar amb més cautela que els altres dos.

Nivell 2 - Anàlisi amb dades enriquides i estructures múltiples
Per a aquest nivell, es facilita informació externa que simula dades reals utilitzades per una empresa en processos d'anàlisi i presa de decisions salarials.

» Fitxer JSON de política salarial: L'empresa disposa d'un fitxer JSON que defineix les regles d'increment salarial anual. Aquest document estableix increments base per grup professional i ajustos addicionals segons factors personals (edat, fills) i contextuals. Aquest tipus d'estructura és habitual en sistemes interns de recursos humans, ja que permet modificar les polítiques sense alterar el codi.
» Fitxer CSV de ciutats europees: Es proporciona també un fitxer CSV amb informació contextual sobre diverses ciutats europees (país, índex de cost de vida, qualitat del transport públic i tipus de clima). Aquest dataset representa una font externa de dades utilitzada per enriquir l'anàlisi salarial en funció de l'entorn geogràfic on resideix cada persona.

Exercici 1: Enriquiment mitjançant datasets externs

🎯 Objectiu: ampliar el dataset principal integrant informació contextual externa i avaluar l'impacte analític de les decisions de fusió.
Aspectes a revisar i analitzar

» Les diferents fonts de dades són coherents i compatibles entre si per poder calcular un salari ajustat?
» La informació contextual incorporada (cost de vida, polítiques salarials) és suficient per aplicar increments de manera consistent?
» Hi ha buits d'informació que impedeixin calcular correctament l'increment salarial en alguns registres?
» És necessari prendre decisions analítiques quan la informació contextual és parcial o absent?
» Com poden aquestes decisions influir en el càlcul del salari ajustat i en la seva interpretació posterior?
Concretament, hauràs de:

» Integrar el dataset principal amb:
El dataset contextual de ciutats.
La política salarial definida en format JSON
» Calcular un increment salarial total combinant:
Grup professional.
Situació familiar.
Franja d'edat.
Cost de vida de la ciutat.
» Identificar quins registres no poden rebre algun dels ajustos i explicar-ne el motiu.
» Decidir, de manera raonada, com tractar aquests casos en el càlcul del salari ajustat, justificando l'impacte analític de la decisió presa.
 Important

L'objectiu no és només enriquir dades, sinó demostrar criteri analític en la gestió de la incertesa i entendre com les decisions d'integració condicionen les anàlisis posteriors.

### Algorisme: Enriquiment del dataset amb dades externes i càlcul de l'increment salarial

1. **Neteja del dataset de ciutats (`ciutats_context.csv`)**: abans d'integrar aquesta font externa, cal netejar-la: eliminar les files completament buides (que actuen com a separadors entre blocs de països), corregir errors de codificació puntuals (com `A Coru\u00f1a` o `Munchen` sense dièresi), i assegurar que els noms de ciutat coincideixen amb els del dataset principal.
2. **Càrrega de la política salarial (`politica_salarial.json`)**: es carrega el fitxer JSON, que conté diversos diccionaris amb els percentatges d'increment segons grup professional, situació familiar, franja d'edat i cost de vida/transport de la ciutat.
3. **Integració (merge) del dataset principal amb el de ciutats**: es creua cada empleat amb la informació de la seva ciutat, utilitzant la combinació de Ciutat i País com a clau (no només el nom de la ciutat), per evitar confondre ciutats homònimes de països diferents (per exemple, "València" a Espanya vs "Valencia" a Veneçuela, o "Paris" a França vs "Paris" als EUA).
4. **Detecció de buits d'informació**: es comprova quins empleats no han trobat cap coincidència amb el dataset de ciutats (per exemple, perquè la seva ciutat no hi apareix), i es documenta quants registres es veuen afectats i per què.
5. **Càlcul de l'increment salarial total**: per a cada empleat, es calcula l'increment combinant:
   - Increment base segons el seu Grup Professional.
   - Ajust segons si té fills o no.
   - Ajust segons la seva franja d'edat.
   - Ajust segons el cost de vida i el transport públic de la seva ciutat (quan aquesta informació estigui disponible).
   
   Es decideix, de manera justificada, com tractar els casos on falti algun d'aquests ajustos (per exemple, aplicar un ajust neutre de 0% si no hi ha dada de ciutat, en comptes de descartar l'empleat).
6. **Aplicació dels límits de la política**: es limita l'increment total al màxim permès (15%, segons `max_total_increment_percent`), i es calcula finalment el nou salari ajustat per a cada empleat.

In [48]:
import json

# Carreguem el dataset de ciutats (separador ';', no coma, si no llegeix com si fos tot una columna)
df_ciutats = pd.read_csv('ciutats_context.csv', sep=';')

# Carreguem la política salarial (fitxer JSON)
with open('politica_salarial.json', 'r', encoding='utf-8') as f:
    politica_salarial = json.load(f)

print(df_ciutats.shape) # 41 files i 5 columnes (estructura)
df_ciutats.head(15) #per a que ens mostri les 15 files primeres

(41, 5)


,Ciutat,País,Cost_vida_index,Transport_public_score,Clima_tipus
0,Barcelona,Espanya,78.0,8.5,Mediterrani
1,Madrid,Espanya,75.0,8.0,Continental
2,València,Espanya,70.0,7.5,Mediterrani
3,Sevilla,Espanya,65.0,7.0,Mediterrani
4,Bilbao,Espanya,72.0,7.8,Oceànic
5,Zaragoza,Espanya,68.0,7.0,Continental
6,Murcia,Espanya,60.0,6.5,Mediterrani
7,Palma,Espanya,73.0,7.2,Mediterrani
8,Girona,Espanya,66.0,6.8,Mediterrani
9,Tarragona,Espanya,67.0,6.9,Mediterrani


In [49]:
# Netegem el dataset de ciutats:
# 1. Eliminem les files completament buides (separadors entre blocs de països)
# 2. Corregim l'escapament Unicode sense processar (\u00f1 -> ñ)
# 3. Corregim "Munchen" (sense dièresi) per fer-lo coincidir amb "München" del dataset principal

files_abans_ciutats = df_ciutats.shape[0]
df_ciutats = df_ciutats.dropna(how='all')
print(f"S'han eliminat {files_abans_ciutats - df_ciutats.shape[0]} files buides (separadors)")

# Corregim l'escapament unicode literal (per exemple "A Coru\u00f1a" -> "A Coruña")
df_ciutats['Ciutat'] = df_ciutats['Ciutat'].apply(
    lambda x: x.encode().decode('unicode_escape') if isinstance(x, str) and '\\u' in x else x
)

# Corregim Munchen -> München perquè coincideixi amb el dataset principal
df_ciutats['Ciutat'] = df_ciutats['Ciutat'].replace('Munchen', 'München')

df_ciutats

S'han eliminat 5 files buides (separadors)


,Ciutat,País,Cost_vida_index,Transport_public_score,Clima_tipus
0,Barcelona,Espanya,78.0,8.5,Mediterrani
1,Madrid,Espanya,75.0,8.0,Continental
2,València,Espanya,70.0,7.5,Mediterrani
3,Sevilla,Espanya,65.0,7.0,Mediterrani
4,Bilbao,Espanya,72.0,7.8,Oceànic
5,Zaragoza,Espanya,68.0,7.0,Continental
6,Murcia,Espanya,60.0,6.5,Mediterrani
7,Palma,Espanya,73.0,7.2,Mediterrani
8,Girona,Espanya,66.0,6.8,Mediterrani
9,Tarragona,Espanya,67.0,6.9,Mediterrani


In [50]:
# Unifiquem el nom de la columna de país per poder fer el merge per Ciutat + País alhora
df_ciutats = df_ciutats.rename(columns={'País': "País d'origen"})

# Fem el merge (left, per no perdre cap empleat encara que no tingui coincidència)
df = df.merge(df_ciutats, on=['Ciutat', "País d'origen"], how='left')

print(df.shape)
df.head()

(905, 20)


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat,Cost_vida_index,Transport_public_score,Clima_tipus
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,1469.0,Grup B,Cap de projecte,3.0,True,32108.000000,25.0,10.1,64.0,6.7,Continental
1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,Home,2718.0,Grup C,Analista senior,0.0,True,19496.000000,10.4,18.7,68.0,7.0,Mediterrani
2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,Dona,1358.0,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7,68.0,7.0,Mediterrani
3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,Home,1478.0,Grup B,Analista,2.0,False,NaN,NaN,18.3,65.0,7.0,Mediterrani
4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,1284.0,Grup B,Administratiu/va,0.0,True,54533.330961,11.8,13.1,72.0,7.8,Oceànic


In [54]:
# Detectem quins empleats no han trobat coincidència amb el dataset de ciutats
# (per exemple, si la seva ciutat no hi apareix registrada)
sense_match = df['Cost_vida_index'].isna()
print(f"Empleats sense informació de ciutat: {sense_match.sum()}")
print(df.loc[sense_match, 'Ciutat'].value_counts())

Empleats sense informació de ciutat: 123
Ciutat
Màlaga        50
Toulouse      14
Nantes        12
Nice          11
Bologna       11
Stavanger      8
Trondheim      6
Tromsø         4
Köln           4
Düsseldorf     3
Name: count, dtype: int64


In [55]:
# Per als empleats la ciutat dels quals no apareix al dataset extern de ciutats,
# imputem el cost de vida i el transport públic amb la mitjana del seu PAÍS,
# ja que és una aproximació més informativa que un ajust neutre (0%),
# tot i assumir que el cost de vida és relativament similar dins del mateix país.
# El clima es deixa com a nul, ja que no té sentit fer una "mitjana" d'un tipus de clima.

mitjana_cost_per_pais = df.groupby("País d'origen")['Cost_vida_index'].transform('mean')
mitjana_transport_per_pais = df.groupby("País d'origen")['Transport_public_score'].transform('mean')

files_imputades = df['Cost_vida_index'].isna().sum()

df['Cost_vida_index'] = df['Cost_vida_index'].fillna(mitjana_cost_per_pais)
df['Transport_public_score'] = df['Transport_public_score'].fillna(mitjana_transport_per_pais)

print(f"S'han imputat {files_imputades} files amb la mitjana de cost de vida i transport del seu país")
print(df['Cost_vida_index'].isna().sum(), "files encara sense dada (si algun país no té cap ciutat coneguda)")

S'han imputat 123 files amb la mitjana de cost de vida i transport del seu país
0 files encara sense dada (si algun país no té cap ciutat coneguda)


### Conclusions

- **Problema detectat:** el dataset extern de ciutats (`ciutats_context.csv`) presentava files buides de separació, errors de codificació puntuals (`A Coru\u00f1a`, `Munchen`) i dues ciutats homònimes de països diferents (Paris a Estats Units, Valencia a Veneçuela) que podrien haver generat encreuaments incorrectes si s'hagués fet el merge només pel nom de la ciutat.
- **Solució aplicada:** es netegen aquests problemes abans d'integrar les dades, i el merge es fa creuant per Ciutat i País d'origen alhora, evitant confondre ciutats homònimes. Un cop integrat, es detecten 123 empleats (13,6% del dataset) la ciutat dels quals no apareixia al dataset extern (per exemple, Màlaga, Toulouse, Nantes, Nice...). Per a aquests casos, s'imputa el cost de vida i el transport públic amb la mitjana del seu país d'origen, ja que ofereix una aproximació més informativa que un ajust neutre, tot i assumir una certa homogeneïtat dins del mateix país.
- **Resultat:** totes les files disposen ara d'informació de cost de vida i transport (0 valors nuls restants), llestes per calcular l'increment salarial.

In [56]:
# Dividim Cost_vida_index i Transport_public_score en 3 trams (low/medium/high)
# utilitzant tercils, ja que la política salarial no especifica els llindars exactes.
df['cost_vida_categoria'] = pd.qcut(df['Cost_vida_index'], q=3, labels=['low', 'medium', 'high'])
df['transport_categoria'] = pd.qcut(df['Transport_public_score'], q=3, labels=['low', 'medium', 'high'])

# Calculem la franja d'edat de cada empleat (any actual - any de naixement)
any_actual = 2026
edat = any_actual - df['Any de Naixement']

def franja_edat(edat):
    if edat < 30:
        return '18-29'
    elif edat < 45:
        return '30-44'
    elif edat < 60:
        return '45-59'
    else:
        return '60+'

df['franja_edat'] = edat.apply(franja_edat)

# Determinem si té fills o no (basat en la columna unificada Nombre_fills)
df['te_fills'] = df['Nombre_fills'] > 0

df[['cost_vida_categoria', 'transport_categoria', 'franja_edat', 'te_fills']].head()

,cost_vida_categoria,transport_categoria,franja_edat,te_fills
0,low,low,60+,True
1,low,low,60+,False
2,low,low,60+,True
3,low,low,30-44,True
4,medium,high,60+,False


In [57]:
# Calculem l'increment salarial total combinant els 4 factors del JSON:
# grup professional, situació familiar, franja d'edat, i cost de vida + transport.

def calcular_increment(fila):
    # 1. Increment base segons Grup Professional
    increment_grup = politica_salarial['base_increment_by_group'].get(fila['Grup Professional'], 0)
    
    # 2. Ajust segons si té fills
    clau_fills = 'has_children' if fila['te_fills'] else 'no_children'
    increment_fills = politica_salarial['children_adjustment'][clau_fills]
    
    # 3. Ajust segons franja d'edat
    increment_edat = politica_salarial['age_adjustment'][fila['franja_edat']]
    
    # 4. Ajust segons cost de vida i transport de la ciutat
    increment_cost_vida = politica_salarial['cost_of_living_adjustment'][fila['cost_vida_categoria']]
    increment_transport = politica_salarial['transport_adjustment'][fila['transport_categoria']]
    
    total = increment_grup + increment_fills + increment_edat + increment_cost_vida + increment_transport
    
    # Apliquem el límit màxim definit a la política
    maxim = politica_salarial['rules']['max_total_increment_percent']
    return min(total, maxim)

df['increment_percent'] = df.apply(calcular_increment, axis=1)

print(df['increment_percent'].describe())

count    905.000000
mean       6.980663
std        1.675537
min        2.500000
25%        6.000000
50%        7.000000
75%        8.000000
max       12.000000
Name: increment_percent, dtype: float64


In [58]:
# Calculem el salari ajustat aplicant l'increment percentual calculat
df['salari_ajustat'] = (df['Salari mensual'] * (1 + df['increment_percent'] / 100)).round(2)

df[['Salari mensual', 'increment_percent', 'salari_ajustat']].head()

,Salari mensual,increment_percent,salari_ajustat
0,1469.0,7.0,1571.83
1,2718.0,4.0,2826.72
2,1358.0,8.5,1473.43
3,1478.0,6.5,1574.07
4,1284.0,5.5,1354.62


### Conclusions

- **Solució aplicada:** es calcula l'increment salarial total combinant quatre factors definits a la política salarial (`politica_salarial.json`): l'increment base segons Grup Professional, l'ajust segons si l'empleat té fills, l'ajust segons la seva franja d'edat, i l'ajust segons el cost de vida i el transport públic de la seva ciutat (dividits en trams low/medium/high mitjançant tercils, ja que la política no especificava els llindars exactes). El resultat es limita al màxim del 15% establert a les regles de la política.
- **Resultat:** l'increment mitjà obtingut és del 6,98% (rang entre 2,5% i 12%, sense arribar mai al límit màxim), i s'aplica sobre el `Salari mensual` per obtenir el `salari_ajustat` final de cada empleat.

Exercici 2: Reestructuració del dataset
🎯 Objectiu: preparar el dataset per a anàlisis comparatives i/o visuals. Convertir informació dispersa en format llarg:
Salaris
Mobilitat
Context climàtic etc
Aspectes a revisar i analitzar

» Existeixen variables que representen conceptes similars o relacionats, però estan distribuïdes en diferents columnes?
»El format actual del dataset facilita o dificulta:
comparacions entre magnituds.
agregacions per grups.
» Quin impacte té la forma del dataset (ample vs llarg) en la lectura i interpretació de les dades?
Concretament, hauràs de:

» Reestructurar el dataset utilitzant melt() per convertir-lo a format llarg, definint prèviament:
les columnes que actuen com a identificadores (per exemple: persona, ciutat, grup professional, gènere, etc.).
les columnes que passen a representar valors mesurats (per exemple: salaris, mobilitat, variables contextuals).
» Analitzar per què el format llarg és més adequat per:
realitzar comparacions entre diferents tipus de variables.
calcular estadístics agregats.
📌 Nota

No es busca generar visualitzacions en aquest punt, sinó preparar correctament les dades per fer-ho en exercicis posteriors.

In [59]:
# Reestructurem el dataset a format llarg (melt), separant les columnes
# identificadores (que descriuen la persona) de les columnes de valors mesurats
# (magnituds numèriques que volem poder comparar entre elles).

id_vars = ['Nom', 'Cognoms', 'DNI', 'Ciutat', "País d'origen", 'Gènere', 'Grup Professional', 'Càrrec', 'franja_edat']

value_vars = [
    'Salari mensual', 'salari_ajustat', 'increment_percent',
    'Km_anuals', 'Consum_mitja_L_100km',
    'Temperatura_mitjana_ciutat', 'Cost_vida_index', 'Transport_public_score'
]

df_llarg = df.melt(
    id_vars=id_vars,
    value_vars=value_vars,
    var_name='variable',
    value_name='valor'
)

print(df_llarg.shape)
df_llarg.head(10)

(7240, 11)


,Nom,Cognoms,DNI,Ciutat,País d'origen,Gènere,Grup Professional,Càrrec,franja_edat,variable,valor
0,Joana,Gil Navarro,66722344X,Valladolid,Espanya,Dona,Grup B,Cap de projecte,60+,Salari mensual,1469.0
1,Marc,Muñoz,48840994W,Alacant,Espanya,Home,Grup C,Analista senior,60+,Salari mensual,2718.0
2,Noa,Serra,14308421X,Alacant,Espanya,Dona,Grup A,Tècnic IT,60+,Salari mensual,1358.0
3,Pol,Gil,58586340F,Sevilla,Espanya,Home,Grup B,Analista,30-44,Salari mensual,1478.0
4,David,Vila,82070937P,Bilbao,Espanya,Dona,Grup B,Administratiu/va,60+,Salari mensual,1284.0
5,Alexia,Puig,25014631B,Las Palmas de Gran Canaria,Espanya,Dona,Grup B,Cap de projecte,45-59,Salari mensual,2069.0
6,David,Gómez Vila,36998038E,Palma,Espanya,Dona,Grup B,Cap de projecte,45-59,Salari mensual,2349.0
7,Carmen,López,47376585K,Barcelona,Espanya,Home,Grup C,Responsable de vendes,60+,Salari mensual,2855.0
8,Nora,Andersen Pedersen,37289414J,Stavanger,Noruega,Dona,Grup B,Analista junior,60+,Salari mensual,1338.0
9,Joana,NaN,82772208R,Valladolid,Espanya,Dona,Grup A,Administratiu/va,60+,Salari mensual,982.0


### Conclusions

- **Transformació aplicada:** el dataset original (format ample, amb una columna per a cada variable) es converteix a format llarg mitjançant `melt()`, mantenint com a identificadors les dades descriptives de cada empleat (nom, ciutat, país, gènere, grup professional, càrrec i franja d'edat), i convertint totes les magnituds numèriques (salaris, mobilitat, context climàtic i de cost de vida) en dues úniques columnes: `variable` (el nom de la magnitud) i `valor` (la seva xifra).
- **Per què el format llarg facilita l'anàlisi:**
  - **Comparacions entre magnituds diferents:** en format ample, comparar per exemple el salari amb el consum de combustible requereix escriure codi específic per a cada parella de columnes. En format llarg, n'hi ha prou amb filtrar per `variable` per aïllar qualsevol magnitud i comparar-la amb qualsevol altra de manera uniforme.
  - **Agregacions per grups:** funcions com `groupby()` combinades amb `variable` permeten calcular estadístics (mitjana, mediana...) de totes les magnituds alhora amb una sola instrucció, en lloc de repetir el càlcul columna per columna com caldria fer en format ample.
  - **Escalabilitat:** si en el futur s'afegeixen noves variables mesurades, el format llarg no necessita cap canvi estructural — simplement apareixeran noves categories dins la columna `variable`, mentre que en format ample caldria afegir noves columnes i adaptar tot el codi d'anàlisi existent.
- **Contrapartida:** el format llarg és menys intuïtiu per a la lectura ràpida d'un registre individual (cal filtrar per identificar totes les dades d'una mateixa persona), per això es manté el dataset ample original per a consultes puntuals, i es reserva el format llarg específicament per a anàlisis comparatives i agregades.

Exercici 3: Creació de variables derivades (Feature Engineering)
🎯 Objectiu: generar noves variables derivades a partir de la informació existent i avaluar-ne la qualitat analítica. Tot i que algunes variables de base (com l’edat o el salari) ja han estat introduïdes en exercicis anteriors, en aquest nivell no es valora la seva creació tècnica, sinó la capacitat de transformar-les en indicadors més informatius, la detecció de casos límit o incoherents i la reflexió sobre la utilitat real de les variables generades en un context analític.
Exemples (tria guiada):

» Índex de dependència familiar: Salari mensual / (Nombre fills + 1).
» Índex de mobilitat: Km anuals / Temperatura mitjana.
» Ràtio salari / cost de vida.
» Indicador de dependència familiar: Salari / Nombre fills.
» Pes de responsabilitat familiar: Nombre fills * Edat.
» Indicador de etapa vital: 0 → sense càrregues, 1 → etapa activa, 2 → alta dependència.
Concretament, hauràs de:

» Crear almenys 3 variables derivades.
» Justificar:
La seva utilitat analítica.
Les decisions preses.
» Detectar casos límit o incoherents generats per les noves variables.
📌 Nota

El feature engineering es realitza sobre el dataset en format ample, ja que la creació de variables derivades requereix combinar informació de naturalesa diferent. El format llarg s'utilitza posteriorment per facilitar comparacions, agregacions i anàlisis descriptives de les variables generades.

### Creació de variables derivades

En aquest apartat es generen tres noves variables a partir de la informació ja disponible al dataset, amb l'objectiu d'obtenir indicadors més informatius que les dades brutes per si soles:

1. **Índex de dependència familiar** (`Salari mensual / (Nombre_fills + 1)`): mesura quants euros de salari corresponen a cada membre de la unitat familiar (l'empleat més els seus fills). Se suma 1 al nombre de fills per evitar dividir per zero quan l'empleat no en té, i perquè l'empleat mateix també forma part de la unitat familiar que ha de mantenir. Un valor baix indica que el salari s'ha de repartir entre més persones; un valor alt indica més marge econòmic per persona.
2. **Ràtio salari / cost de vida** (`salari_ajustat / Cost_vida_index`): el salari per si sol no permet comparar de manera justa el poder adquisitiu real entre persones que viuen en ciutats diferents, ja que un mateix import val molt més en una ciutat barata que en una de cara. Aquesta ràtio aproxima el poder adquisitiu relatiu de cada empleat, tenint en compte l'entorn on viu.
3. **Indicador d'etapa vital** (categòric: 0, 1, 2): resumeix en una única variable la combinació d'edat i càrregues familiars d'un empleat, classificant-lo en:
   - **0 - Sense càrregues:** sense fills i en una franja d'edat jove (18-29 anys).
   - **1 - Etapa activa:** amb responsabilitats moderades (té fills i/o es troba en una franja d'edat intermèdia, 30-59 anys).
   - **2 - Alta dependència:** combinació de factors que suggereixen més càrrega econòmica/familiar (per exemple, tenir un nombre elevat de fills, independentment de l'edat).

Aquestes variables permeten anàlisis posteriors més matisades que les dades originals per separat, com ara identificar perfils d'empleats amb menys marge econòmic relatiu o comparar el poder adquisitiu real entre ciutats/països.

In [60]:
# 1. Índex de dependència familiar: salari repartit entre l'empleat i els seus fills.
# Sumem 1 al nombre de fills perquè l'empleat mateix també forma part de la unitat
# familiar, i per evitar dividir per zero quan no té fills.
df['index_dependencia_familiar'] = (df['Salari mensual'] / (df['Nombre_fills'] + 1)).round(2)

# 2. Ràtio salari / cost de vida: aproxima el poder adquisitiu relatiu de l'empleat
# tenint en compte l'entorn on viu (una ciutat més cara "consumeix" més el salari).
df['ratio_salari_cost_vida'] = (df['salari_ajustat'] / df['Cost_vida_index']).round(2)

# 3. Indicador d'etapa vital: combina franja d'edat i nombre de fills en una
# única categoria (0=sense càrregues, 1=etapa activa, 2=alta dependència).
def etapa_vital(fila):
    if fila['Nombre_fills'] >= 3:
        return 2  # alta dependència: nombre elevat de fills
    elif fila['franja_edat'] == '18-29' and fila['Nombre_fills'] == 0:
        return 0  # sense càrregues: jove i sense fills
    else:
        return 1  # etapa activa: la resta de casos

df['etapa_vital'] = df.apply(etapa_vital, axis=1)

print(df[['index_dependencia_familiar', 'ratio_salari_cost_vida', 'etapa_vital']].describe())
print(df['etapa_vital'].value_counts())

       index_dependencia_familiar  ratio_salari_cost_vida  etapa_vital
count                  877.000000              905.000000   905.000000
mean                  1145.517685               23.852917     1.047514
std                    675.905314                9.721290     0.494390
min                    143.600000                9.280000     0.000000
25%                    617.000000               16.580000     1.000000
50%                   1067.000000               21.470000     1.000000
75%                   1524.000000               28.370000     1.000000
max                   3672.000000               65.440000     2.000000
etapa_vital
1    682
2    133
0     90
Name: count, dtype: int64


In [61]:
# Cas límit 1: persones sense Nombre_fills (per tant, sense índex de dependència calculable)
print("Files amb Nombre_fills nul:", df['Nombre_fills'].isna().sum())

# Cas límit 2: valor màxim de la ràtio salari/cost de vida
print(df.nlargest(3, 'ratio_salari_cost_vida')[['Salari mensual', 'Cost_vida_index', 'ratio_salari_cost_vida']])

Files amb Nombre_fills nul: 28
     Salari mensual  Cost_vida_index  ratio_salari_cost_vida
11           3672.0             62.0                   65.44
638          3437.0             60.0                   63.87
763          3425.0             65.0                   59.02


### Conclusions — Detecció de casos límit i incoherències en les variables derivades

- **`index_dependencia_familiar`:** presenta 28 valors nuls, corresponents a empleats sense cap informació disponible sobre el nombre de fills (ni `Fills` ni `No Fills` tenien valor a l'origen, i per tant `Nombre_fills` va quedar buit després de la unificació de columnes). Aquests casos no es poden calcular ni s'han d'imputar arbitràriament, ja que assumir un nombre de fills sense cap indici seria inventar informació.
- **`ratio_salari_cost_vida`:** els valors més alts (fins a 65,4) corresponen a empleats amb salaris molt elevats (3.400-3.700 €) que resideixen en ciutats amb un cost de vida baix (60-65 punts). No es tracta d'errors, sinó de casos reals i coherents: un salari alt en un entorn econòmicament assequible es tradueix lògicament en un poder adquisitiu relatiu molt superior a la mitjana.
- **`etapa_vital`:** la distribució (90 en etapa 0, 682 en etapa 1, 133 en etapa 2) sembla raonable i sense valors inesperats, ja que la majoria d'empleats es troben en l'etapa activa intermèdia.

Exercici 4: Exportació i síntesi de resultats

🎯 Objectiu: consolidar els resultats obtinguts després de la preparació, enriquiment i creació de variables, i generar sortides estructurades reutilitzables per a anàlisi o reporting.
En aquest exercici es vol posar el focus en:

» L'organització de les dades per grups rellevants.
» La generació de datasets finals coherents.
» La capacitat de sintetitzar informació clau.
Aspectes a revisar i analitzar:

» Quin nivell d'agrupació és més adequat per a l'exportació de resultats?
» Com separar dades detallades de dades agregades?
» Quina informació és rellevant per a un ús posterior (anàlisi, visualització, reporting)?
Concretament, hauràs de:

» Utilitzant un bucle, exportar en 4 fitxers (format .csv o .xlsx) les dades corresponents a cada Grup Professional. Per exemple: dades_GrupA.csv, dades_GrupB.csv...
» Crear i exportar un cinquè DataFrame (format .csv o .xlsx) que contingui, per a cada Grup Professional:
Nombre total de treballadors.
Salari mitjà.
Edat mediana.
(EXTRA)
Explora com exportar tots els DataFrames en un únic fitxer Excel, utilitzant:

» Una pestanya per a cada Grup Professional.
» Una pestanya addicional amb el resum agregat.
📌 Nota

No es valora la complexitat tècnica, sinó la claredat, l'ordre i la coherència de les sortides generades.

In [62]:
# Pas 1: Calculem l'edat numèrica de cada empleat (necessària per a l'edat mediana del resum)
any_actual = 2026
df['Edat'] = any_actual - df['Any de Naixement']

In [63]:
# Pas 2: Exportem un fitxer CSV per a cada Grup Professional utilitzant un bucle,
# en comptes d'escriure el codi manualment per a cada grup per separat.
for grup in df['Grup Professional'].unique():
    df_grup = df[df['Grup Professional'] == grup]
    nom_fitxer = f"dades_{grup.replace(' ', '')}.csv"
    df_grup.to_csv(nom_fitxer, index=False)
    print(f"Exportat: {nom_fitxer} ({df_grup.shape[0]} files)")

Exportat: dades_GrupB.csv (321 files)
Exportat: dades_GrupC.csv (130 files)
Exportat: dades_GrupA.csv (416 files)
Exportat: dades_GrupD.csv (38 files)


In [64]:
# Pas 3: Creem un DataFrame resum amb estadístiques clau per Grup Professional:
# nombre de treballadors, salari mitjà i edat mediana.
resum_grups = df.groupby('Grup Professional').agg(
    nombre_treballadors=('DNI', 'count'),
    salari_mitja=('Salari mensual', 'mean'),
    edat_mediana=('Edat', 'median')
).round(2)

resum_grups.to_csv('resum_grups_professionals.csv')
resum_grups

,nombre_treballadors,salari_mitja,edat_mediana
Grup Professional,,,
Grup A,416,1120.53,44.0
Grup B,321,1650.58,45.0
Grup C,130,2352.47,48.0
Grup D,38,3181.53,39.0


In [65]:
# (EXTRA) Consolidem totes les taules en un únic fitxer Excel,
# amb una pestanya per a cada Grup Professional i una pestanya addicional
# amb el resum agregat, per facilitar la consulta conjunta.

with pd.ExcelWriter('dades_per_grups.xlsx', engine='openpyxl') as writer:
    for grup in df['Grup Professional'].unique():
        df_grup = df[df['Grup Professional'] == grup]
        nom_pestanya = grup.replace(' ', '_')  # els noms de pestanya no poden tenir espais problemàtics
        df_grup.to_excel(writer, sheet_name=nom_pestanya, index=False)

    resum_grups.to_excel(writer, sheet_name='Resum')

print("Fitxer 'dades_per_grups.xlsx' creat amb totes les pestanyes")

Fitxer 'dades_per_grups.xlsx' creat amb totes les pestanyes


### Conclusions

- **Dades detallades:** s'exporten 4 fitxers CSV independents (`dades_GrupA.csv`, `dades_GrupB.csv`, `dades_GrupC.csv`, `dades_GrupD.csv`), un per cada Grup Professional, mitjançant un bucle que recorre automàticament tots els grups sense necessitat de repetir codi.
- **Dades agregades:** es genera un cinquè fitxer (`resum_grups_professionals.csv`) amb el nombre de treballadors, el salari mitjà i l'edat mediana de cada grup, separant clarament la informació detallada (una fila per empleat) de la informació sintetitzada (una fila per grup).
- **Consolidació (extra):** totes les taules es combinen en un únic fitxer Excel (`dades_per_grups.xlsx`), amb una pestanya per a cada Grup Professional i una pestanya addicional (`Resum`) amb les estadístiques agregades.
- **Observació clau del resum:** s'aprecia una relació inversa entre mida del grup i salari mitjà — el **Grup D** té el nombre més baix de treballadors (38) però el salari mitjà més alt (3.181,53 €), mentre que el **Grup A** és el més nombrós (416 treballadors) amb el salari mitjà més baix (1.120,53 €).